daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -279,6 +279,18 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
 
   def roots: ImmArray[Nid]
 
+  // TODO: https://github.com/digital-asset/daml/issues/8020
+  //  for now we assume that rollback node cannot be a root of a transaction.
+  def rootNodes: ImmArray[Node.GenActionNode[Nid, Cid]] =
+    roots.map(nid =>
+      nodes(nid) match {
+        case action: Node.GenActionNode[Nid, Cid] =>
+          action
+        case _: Node.NodeRollback[_] =>
+          sys.error("illed-formed transaction")
```

Comment:
```suggestion
          sys.error("ill-formed transaction")
```
Assuming my comment above is correct, we should also throw something sensible like an IllegalArgumentException here instead of `sys.error`.

Reply:
:+1: